# StructureFold2 <img src='assets/sf2_logo.png' align='right' width='400px' />

For information on using StructureFold2, Please consult the included manual, SF2_Manual.pdf.
If you use StructureFold2, please cite [_Tack et al._, 2018](https://www.sciencedirect.com/science/article/pii/S1046202317303535).


**Software Dependencies**
+ [Python 2.7.X](https://www.python.org/)
+ [BioPython](https://biopython.org/)
+ [Numpy](https://numpy.org/)
+ [Cutadapt](https://cutadapt.readthedocs.io/en/stable/)
+ [SAMtools](http://samtools.sourceforge.net/)
+ [Bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)

**Recommended Software**
+ [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
+ [RNAStructure](https://rna.urmc.rochester.edu/RNAstructure.html)
+ [R](https://www.r-project.org/)

## Updates and Errata

### structure_statistics.py
structure_statisics.py replaces make_standed_csv.py and make_MFE_csv.py, with the added
functionality of being able to combine multiple directories of <.ct> files (connectivity tables) generated by batch_fold_rna.py
into a single report. Older versions of RNAStructure did not include a DeltaG value for folds performed under MFE settings, 
but the new more consistent reporting of <.ct> files allows both strandedness and DeltaG to be extracted from such 
MFE folds, thus making one module sufficient for both tasks. Fused mode (-mode F) summarizes one or multiple 
directories, generating a single <.csv> organized by transcript. When using multiple 
directories, the names of the directories must be different. If your transcript names contain underscores, 
input the number they contain (-offset). Raw mode (-mode R) instead operates on a single directory and 
generates a single <.csv> organized by the files contained in the directory, and may be of use when 
it is hard to automatically parse transcript names.

#### Usage
```
Summarizes MFE <.ct> files

optional arguments:
  -h, --help      show this help message and exit
  -d D [D ...]    CT directory/directories
  -mode {F,R}     Fused/Raw statistics
  -name NAME      Output file name
  -offset OFFSET  Number of Underscores in Transcript Names
```

## Planned Updates

* make_PPV_csv.py will be merged into structure_statisics.py
* coverage_overlap.py will be merged into rtsc_coverage.py
* Features that are exclusive to <.react> files have been requested to be available for <.rtsc>,
or vice-versa. I.E, rather than making a separate module for react_correlation to complement rtsc_correlation,
both features will be included in a single module; modules that work on both will be prefixed with 'rx'. So react_correlation and rtsc_correlation
would become rx_correlation, thus deprecating the old module(s) entirely.

